pkg/cephfs/volume.go

```diff
@@ -46,111 +36,27 @@ func getVolumeNamespace(volID volumeID) string {
 }
 
-func setVolumeAttribute(root, attrName, attrValue string) error {
-	return execCommandErr("setfattr", "-n", attrName, "-v", attrValue, root)
-}
-
-func createVolume(volOptions *volumeOptions, adminCr *credentials, volID volumeID, bytesQuota int64) error {
-	if err := mountCephRoot(volID, volOptions, adminCr); err != nil {
-		return err
-	}
-	defer unmountCephRoot(volID)
-
-	var (
-		volRoot         = getCephRootVolumePathLocal(volID)
-		volRootCreating = volRoot + "-creating"
-	)
-
-	if pathExists(volRoot) {
-		klog.V(4).Infof("cephfs: volume %s already exists, skipping creation", volID)
-		return nil
-	}
-
-	if err := createMountPoint(volRootCreating); err != nil {
-		return err
-	}
-
-	if bytesQuota > 0 {
-		if err := setVolumeAttribute(volRootCreating, "ceph.quota.max_bytes", fmt.Sprintf("%d", bytesQuota)); err != nil {
-			return err
-		}
-	}
-
-	if err := setVolumeAttribute(volRootCreating, "ceph.dir.layout.pool", volOptions.Pool); err != nil {
-		return fmt.Errorf("%v\ncephfs: Does pool '%s' exist?", err, volOptions.Pool)
-	}
-
-	if err := setVolumeAttribute(volRootCreating, "ceph.dir.layout.pool_namespace", getVolumeNamespace(volID)); err != nil {
-		return err
-	}
-
-	if err := os.Rename(volRootCreating, volRoot); err != nil {
-		return fmt.Errorf("couldn't mark volume %s as created: %v", volID, err)
-	}
-
-	return nil
-}
-
-func purgeVolume(volID volumeID, adminCr *credentials, volOptions *volumeOptions) error {
-	if err := mountCephRoot(volID, volOptions, adminCr); err != nil {
-		return err
-	}
-	defer unmountCephRoot(volID)
-
-	var (
-		volRoot         = getCephRootVolumePathLocal(volID)
-		volRootDeleting = volRoot + "-deleting"
-	)
-
-	if pathExists(volRoot) {
-		if err := os.Rename(volRoot, volRootDeleting); err != nil {
-			return fmt.Errorf("couldn't mark volume %s for deletion: %v", volID, err)
-		}
-	} else {
-		if !pathExists(volRootDeleting) {
-			klog.V(4).Infof("cephfs: volume %s not found, assuming it to be already deleted", volID)
-			return nil
-		}
-	}
-
-	if err := os.RemoveAll(volRootDeleting); err != nil {
-		return fmt.Errorf("failed to delete volume %s: %v", volID, err)
-	}
-
-	return nil
-}
-
-func mountCephRoot(volID volumeID, volOptions *volumeOptions, adminCr *credentials) error {
-	cephRoot := getCephRootPathLocal(volID)
-
-	// Root path is not set for dynamically provisioned volumes
-	// Access to cephfs's / is required
-	volOptions.RootPath = "/"
-
-	if err := createMountPoint(cephRoot); err != nil {
-		return err
-	}
-
-	m, err := newMounter(volOptions)
-	if err != nil {
-		return fmt.Errorf("failed to create mounter: %v", err)
-	}
-
-	if err = m.mount(cephRoot, adminCr, volOptions); err != nil {
-		return fmt.Errorf("error mounting ceph root: %v", err)
-	}
+func createVolume(volOptions *volumeOptions, volID volumeID, bytesQuota int64) error {
+	_, _, err := util.ExecCommand(
+		"ceph",
+		"fs",
+		"subvolume",
+		"create",
+		volOptions.FsName,
+		string(volID),
+		string(bytesQuota))
```

Comment:
@poornimag To allow non root write to the directory we could make use of `mode` arg in this call and set mode 777 and root:root instead of 755

Reply:
@humblec, there is no 'mode' args in the `ceph fs subvolume create` command. If that's what you meant.

Reply:
@ajarr Oh.. I noticed this in one of the PR:
>
def create_subvolume(self, subvolume_path, size=None, namespace_isolated=True, mode=0o755):
>

which list `mode` arg.

Reply:
That's the method that's called by the command `fs subvolume create`,
https://github.com/ceph/ceph/blob/master/src/pybind/mgr/volumes/module.py#L353
The 'mode' arg is not exposed to the command line user. Is this required for CSI?

Reply:
@ajarr thats required. Because the app pod can be run as any user and this user should have access to this volume by exposing the mode as mentioned here https://github.com/ceph/ceph-csi/pull/400#discussion_r291903118. If not, the pod user will get `Permission denied` and the volume is not usable.

Reply:
How was  @poornimag able to write files to the pod? https://github.com/ceph/ceph-csi/pull/400#issuecomment-500104579  

@poornimag, ?

OK. We can allow the perms of the directory to be changed. You wanted the owner:group to be changed to root:root. What's it right now?

Reply:
@ajarr In above example, the write happened as 'root' user, thats why she was able to write to it. However as mentioned in earlier comment, pod can be run as any user, so we should allow the write to the volume. We could work around it with the `mode` or setting `777`. iic, `owner:group` is (already) `root:root` now. We only need change in `mode`. @poornimag can confirm.

Reply:
https://github.com/ceph/ceph-csi/pull/400#issuecomment-500677369

Reply:
@humblec setting `0777` on `targetPath` will help
@poornimag  can you try once?

create targetPath with `0777` and once the mounting is done and also change the permission of targetPath to `0777` at last

Reply:
I tested with the pod template you mentioned, but it didn't run. Hence built my own. The volume is not accessible by a non root user:
```
/ $ ls /target/
ls: can't open '/target/': Permission denied
/ $ 
```
If we set the targetpath chmod to 777 during mount in the nodeserver, it will work, but thats doing it everytime we mount, or are you suggesting during the create volume? @Madhu-1 

```[root@Centos1 alpine-nonroot]# cat Dockerfile 
FROM alpine:latest

# Create a group and user
RUN addgroup -g 1015 testgroup5 && adduser -S testuser5 -u 1015 -g 1015
#
# # Tell docker that all future commands should run as the appuser user
USER 1015


[root@Centos1 ceph]# cat test-non-root-user.yaml 
apiVersion: v1
kind: Pod
metadata:
  name: pod-run-as-non-root
spec:
  containers:
  - name: main
    image: alpine-non-root:latest
    command: ["/bin/sleep", "999999"]
    securityContext:
       runAsNonRoot: true
    imagePullPolicy: "IfNotPresent"
    volumeMounts:
     - name: target
       mountPath: /target
  volumes:
   - name: target
     persistentVolumeClaim:
       claimName: fs-perf-target

```

Reply:
@poornimag @Madhu-1 as mentioned earlier (https://github.com/ceph/ceph-csi/pull/400#discussion_r291903118) its better to do it at create time and it should be sufficient. Its just that , the `mode` has to be exposed in the createVolume() func as mentioned earlier and iic, @ajarr is working on this or will be on it. We will use it once available. For now, as a workaround we can set it on targetPath.

Reply:
@poornimag changing permission after the mount is good.